File: src/lib/doctor/service.rs
use crate::doctor::model::{
    DiagnosticReport, DoctorError, FixReport, IssueType, Severity, ValidationIssue,
};
use crate::doctor::ports::DoctorService;
use crate::subtitle::ports::SubtitleService;
use crate::subtitle::service::SubRipService;
use log::{debug, error, info};
use std::fs;
use std::path::PathBuf;

/// SubRip (.srt) doctor service implementation
pub struct SubRipDoctorService {
    base_dir: PathBuf,
}

/// Parser state machine for line-by-line analysis
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum ParserState {
    ExpectingTimestamp,
    ExpectingText,
    BetweenBlocks,
}

impl SubRipDoctorService {
    /// Create a new SubRip doctor service
    ///
    /// # Arguments
    /// * `base_dir` - Base directory for resolving relative file paths
    pub fn new(base_dir: PathBuf) -> Self {
        Self { base_dir }
    }

    /// Validate filename for path traversal attacks
    fn validate_filename(&self, filename: &str) -> Result<(), DoctorError> {
        if filename.is_empty() {
            return Err(DoctorError::InvalidPath(
                "filename cannot be empty".to_string(),
            ));
        }

        if filename.contains("..") || filename.contains('/') || filename.contains('\\') {
            return Err(DoctorError::InvalidPath(format!(
                "invalid filename (path traversal detected): {}",
                filename
            )));
        }

        Ok(())
    }

    /// Get full path to file
    fn get_file_path(&self, filename: &str) -> Result<PathBuf, DoctorError> {
        self.validate_filename(filename)?;
        let path = self.base_dir.join(filename);

        if !path.exists() {
            return Err(DoctorError::FileNotFound(format!("{}", path.display())));
        }

        Ok(path)
    }

    /// Parse a timestamp in HH:MM:SS,mmm format
    fn parse_timestamp(&self, timestamp_str: &str) -> Result<u64, String> {
        let parts: Vec<&str> = timestamp_str.split(&[':', ',']).collect();
        if parts.len() != 4 {
            return Err(format!(
                "expected format HH:MM:SS,mmm, got {}",
                timestamp_str
            ));
        }

        let hours: u64 = parts[0]
            .parse()
            .map_err(|_| format!("invalid hours: {}", parts[0]))?;
        let minutes: u64 = parts[1]
            .parse()
            .map_err(|_| format!("invalid minutes: {}", parts[1]))?;
        let seconds: u64 = parts[2]
            .parse()
            .map_err(|_| format!("invalid seconds: {}", parts[2]))?;
        let millis: u64 = parts[3]
            .parse()
            .map_err(|_| format!("invalid milliseconds: {}", parts[3]))?;

        if minutes >= 60 {
            return Err(format!("minutes must be < 60, got {}", minutes));
        }
        if seconds >= 60 {
            return Err(format!("seconds must be < 60, got {}", seconds));
        }
        if millis >= 1000 {
            return Err(format!("milliseconds must be < 1000, got {}", millis));
        }

        let total_ms = hours * 3600000 + minutes * 60000 + seconds * 1000 + millis;
        Ok(total_ms)
    }
}

impl DoctorService for SubRipDoctorService {
    fn diagnose(&self, filename: &str) -> Result<DiagnosticReport, DoctorError> {
        info!("diagnostics for file '{}'", filename);

        let file_path = self.get_file_path(filename)?;
        let content = fs::read_to_string(&file_path)?;
        let lines: Vec<&str> = content.lines().collect();
        let total_lines = lines.len();

        debug!("file contain lines: {}", total_lines);

        let mut issues = Vec::new();
        let mut state = ParserState::BetweenBlocks;
        let mut current_block = 0;
        let mut empty_line_count = 0;
        let mut text_collected = false;
        let mut consecutive_empty_start = 0;

        for (line_num, line) in lines.iter().enumerate() {
            let line_number = line_num + 1; // 1-based
            let is_empty = line.trim().is_empty();

            match state {
                ParserState::BetweenBlocks => {
                    if is_empty {
                        if empty_line_count == 0 {
                            consecutive_empty_start = line_number;
                        }
                        empty_line_count += 1;
                    } else {
                        if empty_line_count > 1 {
                            issues.push(ValidationIssue {
                                line_number: consecutive_empty_start,
                                block_number: None,
                                issue_type: IssueType::MultipleEmptyLines {
                                    count: empty_line_count,
                                },
                                severity: Severity::Warning,
                                context: Some(format!(
                                    "lines {}-{}",
                                    consecutive_empty_start,
                                    consecutive_empty_start + empty_line_count - 1
                                )),
                            });
                        }
                        empty_line_count = 0;
                        current_block += 1;
                        text_collected = false;

                        if line.trim().parse::<u32>().is_err() {
                            issues.push(ValidationIssue {
                                line_number,
                                block_number: Some(current_block),
                                issue_type: IssueType::InvalidIndex {
                                    value: line.to_string(),
                                    reason: "not a valid number".to_string(),
                                },
                                severity: Severity::Error,
                                context: None,
                            });
                        } else if let Ok(idx) = line.trim().parse::<u32>()
                            && idx < 1
                        {
                            issues.push(ValidationIssue {
                                line_number,
                                block_number: Some(current_block),
                                issue_type: IssueType::InvalidIndex {
                                    value: line.to_string(),
                                    reason: "must be >= 1".to_string(),
                                },
                                severity: Severity::Error,
                                context: None,
                            });
                        }
                        state = ParserState::ExpectingTimestamp;
                    }
                }
                ParserState::ExpectingTimestamp => {
                    if is_empty {
                        issues.push(ValidationIssue {
                            line_number,
                            block_number: Some(current_block),
                            issue_type: IssueType::EmptyLineInBlock,
                            severity: Severity::Error,
                            context: Some("expected timestamp line".to_string()),
                        });
                        state = ParserState::BetweenBlocks;
                    } else {
                        if !line.contains(" --> ") {
                            issues.push(ValidationIssue {
                                line_number,
                                block_number: Some(current_block),
                                issue_type: IssueType::MalformedTimestamp {
                                    value: line.to_string(),
                                    reason: "missing ' --> ' separator".to_string(),
                                },
                                severity: Severity::Error,
                                context: None,
                            });
                        } else {
                            let parts: Vec<&str> = line.split(" --> ").collect();
                            if parts.len() == 2 {
                                match (
                                    self.parse_timestamp(parts[0]),
                                    self.parse_timestamp(parts[1]),
                                ) {
                                    (Ok(start_ms), Ok(end_ms)) => {
                                        if end_ms <= start_ms {
                                            issues.push(ValidationIssue {
                                                line_number,
                                                block_number: Some(current_block),
                                                issue_type: IssueType::InvalidTimestamp {
                                                    start: parts[0].to_string(),
                                                    end: parts[1].to_string(),
                                                },
                                                severity: Severity::Error,
                                                context: None,
                                            });
                                        }
                                    }
                                    (Err(e), _) => {
                                        issues.push(ValidationIssue {
                                            line_number,
                                            block_number: Some(current_block),
                                            issue_type: IssueType::MalformedTimestamp {
                                                value: parts[0].to_string(),
                                                reason: e,
                                            },
                                            severity: Severity::Error,
                                            context: None,
                                        });
                                    }
                                    (_, Err(e)) => {
                                        issues.push(ValidationIssue {
                                            line_number,
                                            block_number: Some(current_block),
                                            issue_type: IssueType::MalformedTimestamp {
                                                value: parts[1].to_string(),
                                                reason: e,
                                            },
                                            severity: Severity::Error,
                                            context: None,
                                        });
                                    }
                                }
                            }
                        }
                        state = ParserState::ExpectingText;
                    }
                }
                ParserState::ExpectingText => {
                    if is_empty {
                        if !text_collected {
                            issues.push(ValidationIssue {
                                line_number,
                                block_number: Some(current_block),
                                issue_type: IssueType::EmptyText,
                                severity: Severity::Error,
                                context: Some("no text before empty line".to_string()),
                            });
                        }
                        state = ParserState::BetweenBlocks;
                        empty_line_count = 0;
                    } else {
                        text_collected = true;
                    }
                }
            }
        }

        if state == ParserState::BetweenBlocks && empty_line_count > 1 {
            issues.push(ValidationIssue {
                line_number: consecutive_empty_start,
                block_number: None,
                issue_type: IssueType::MultipleEmptyLines {
                    count: empty_line_count,
                },
                severity: Severity::Warning,
                context: Some("trailing empty lines".to_string()),
            });
        }

        debug!("detected problematic lines: {}", issues.len());

        let subtitle_service = SubRipService::new(self.base_dir.clone());
        let is_parsable = subtitle_service.get_all(filename).is_ok();

        info!(
            "diagnostics completed: found {} errors, {} warnings",
            issues
                .iter()
                .filter(|i| i.severity == Severity::Error)
                .count(),
            issues
                .iter()
                .filter(|i| i.severity == Severity::Warning)
                .count()
        );

        Ok(DiagnosticReport {
            file_path: file_path.display().to_string(),
            total_lines,
            total_blocks: current_block,
            issues,
            is_parsable,
        })
    }

    fn fix(&self, filename: &str) -> Result<FixReport, DoctorError> {
        info!("starting file fix: {}", filename);

        let file_path = self.get_file_path(filename)?;

        let content = fs::read_to_string(&file_path)?;
        let lines: Vec<&str> = content.lines().collect();

        let mut output_lines = Vec::new();
        let mut state = ParserState::BetweenBlocks;
        let mut current_block_lines = Vec::new();
        let mut empty_line_count = 0;
        let mut issues_fixed = 0;
        let unfixable_issues = Vec::new();

        for line in lines.iter() {
            let is_empty = line.trim().is_empty();

            match state {
                ParserState::BetweenBlocks => {
                    if is_empty {
                        empty_line_count += 1;
                    } else {
                        if !output_lines.is_empty() {
                            if empty_line_count > 1 {
                                issues_fixed += 1;
                            }
                            output_lines.push(String::new());
                        }
                        empty_line_count = 0;
                        current_block_lines.clear();
                        current_block_lines.push(line.to_string());
                        state = ParserState::ExpectingTimestamp;
                    }
                }
                ParserState::ExpectingTimestamp => {
                    if is_empty {
                        issues_fixed += 1;
                    } else {
                        current_block_lines.push(line.to_string());
                        state = ParserState::ExpectingText;
                    }
                }
                ParserState::ExpectingText => {
                    if is_empty {
                        // End of block - write collected lines
                        for block_line in &current_block_lines {
                            output_lines.push(block_line.clone());
                        }
                        state = ParserState::BetweenBlocks;
                        empty_line_count = 0;
                    } else {
                        current_block_lines.push(line.to_string());
                    }
                }
            }
        }

        if !current_block_lines.is_empty() {
            for block_line in &current_block_lines {
                output_lines.push(block_line.clone());
            }
        }

        let fixed_content = output_lines.join("\n");
        fs::write(&file_path, fixed_content)?;

        debug!("fixed file written");

        let subtitle_service = SubRipService::new(self.base_dir.clone());
        let validation_success = subtitle_service.get_all(filename).is_ok();

        if validation_success {
            info!("validation successful: file can be parsed");
        } else {
            error!("validation failed: file still contains errors");
        }

        info!("fix completed: {} issues fixed", issues_fixed);

        Ok(FixReport {
            original_path: file_path.display().to_string(),
            fixed_path: file_path.display().to_string(),
            issues_fixed,
            issues_unfixable: unfixable_issues.len(),
            unfixable_issues,
            validation_success,
        })
    }
}

File: pkg/doctor/service.go
package doctor

import (
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// SubRipDoctorService is the SubRip (.srt) doctor service implementation.
// Was Rust struct SubRipDoctorService.
type SubRipDoctorService struct {
	baseDir string
}

// parserState is the parser state machine for line-by-line analysis.
// Was Rust enum ParserState.
type parserState int

const (
	expectingTimestamp parserState = iota
	expectingText
	betweenBlocks
)

// NewSubRipDoctorService creates a new SubRip doctor service.
//
// baseDir is the base directory for resolving relative file paths.
// Was SubRipDoctorService::new.
func NewSubRipDoctorService(baseDir string) *SubRipDoctorService {
	return &SubRipDoctorService{baseDir: baseDir}
}

// validateFilename validates a filename for path traversal attacks.
// Was SubRipDoctorService::validate_filename.
func (s *SubRipDoctorService) validateFilename(filename string) error {
	if filename == "" {
		return NewInvalidPath("filename cannot be empty")
	}

	if strings.Contains(filename, "..") || strings.Contains(filename, "/") || strings.Contains(filename, "\\") {
		return NewInvalidPath(fmt.Sprintf("invalid filename (path traversal detected): %s", filename))
	}

	return nil
}

// getFilePath gets the full path to the file.
// Was SubRipDoctorService::get_file_path.
func (s *SubRipDoctorService) getFilePath(filename string) (string, error) {
	if err := s.validateFilename(filename); err != nil {
		return "", err
	}

	path := filepath.Join(s.baseDir, filename)

	if !pathExists(path) {
		return "", NewFileNotFound(path)
	}

	return path, nil
}

// parseTimestamp parses a timestamp in HH:MM:SS,mmm format.
// Was SubRipDoctorService::parse_timestamp. On error it returns the same
// message strings as the Rust impl.
func (s *SubRipDoctorService) parseTimestamp(timestampStr string) (uint64, error) {
	parts := splitOnSeparators(timestampStr, ':', ',')
	if len(parts) != 4 {
		return 0, fmt.Errorf("expected format HH:MM:SS,mmm, got %s", timestampStr)
	}

	hours, err := strconv.ParseUint(parts[0], 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid hours: %s", parts[0])
	}
	minutes, err := strconv.ParseUint(parts[1], 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid minutes: %s", parts[1])
	}
	seconds, err := strconv.ParseUint(parts[2], 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid seconds: %s", parts[2])
	}
	millis, err := strconv.ParseUint(parts[3], 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid milliseconds: %s", parts[3])
	}

	if minutes >= 60 {
		return 0, fmt.Errorf("minutes must be < 60, got %d", minutes)
	}
	if seconds >= 60 {
		return 0, fmt.Errorf("seconds must be < 60, got %d", seconds)
	}
	if millis >= 1000 {
		return 0, fmt.Errorf("milliseconds must be < 1000, got %d", millis)
	}

	totalMs := hours*3600000 + minutes*60000 + seconds*1000 + millis
	return totalMs, nil
}

// Diagnose diagnoses a subtitle file for issues.
// Was SubRipDoctorService::diagnose.
func (s *SubRipDoctorService) Diagnose(filename string) (DiagnosticReport, error) {
	slog.Info(fmt.Sprintf("diagnostics for file '%s'", filename))

	filePath, err := s.getFilePath(filename)
	if err != nil {
		return DiagnosticReport{}, err
	}

	content, err := os.ReadFile(filePath)
	if err != nil {
		return DiagnosticReport{}, NewIoError(err)
	}

	lines := splitLines(string(content))
	totalLines := len(lines)

	slog.Debug(fmt.Sprintf("file contain lines: %d", totalLines))

	var issues []ValidationIssue
	state := betweenBlocks
	currentBlock := 0
	emptyLineCount := 0
	textCollected := false
	consecutiveEmptyStart := 0

	for lineNum, line := range lines {
		lineNumber := lineNum + 1 // 1-based
		isEmpty := rustTrim(line) == ""

		switch state {
		case betweenBlocks:
			if isEmpty {
				if emptyLineCount == 0 {
					consecutiveEmptyStart = lineNumber
				}
				emptyLineCount++
			} else {
				if emptyLineCount > 1 {
					ctx := fmt.Sprintf("lines %d-%d", consecutiveEmptyStart, consecutiveEmptyStart+emptyLineCount-1)
					issues = append(issues, ValidationIssue{
						LineNumber:  consecutiveEmptyStart,
						BlockNumber: nil,
						IssueType: IssueType{
							Kind:  MultipleEmptyLines,
							Count: emptyLineCount,
						},
						Severity: Warning,
						Context:  &ctx,
					})
				}
				emptyLineCount = 0
				currentBlock++
				textCollected = false

				blockNum := currentBlock
				idx, parseErr := parseU32(rustTrim(line))
				if parseErr != nil {
					issues = append(issues, ValidationIssue{
						LineNumber:  lineNumber,
						BlockNumber: &blockNum,
						IssueType: IssueType{
							Kind:   InvalidIndex,
							Value:  line,
							Reason: "not a valid number",
						},
						Severity: Error,
						Context:  nil,
					})
				} else if idx < 1 {
					issues = append(issues, ValidationIssue{
						LineNumber:  lineNumber,
						BlockNumber: &blockNum,
						IssueType: IssueType{
							Kind:   InvalidIndex,
							Value:  line,
							Reason: "must be >= 1",
						},
						Severity: Error,
						Context:  nil,
					})
				}
				state = expectingTimestamp
			}
		case expectingTimestamp:
			if isEmpty {
				blockNum := currentBlock
				ctx := "expected timestamp line"
				issues = append(issues, ValidationIssue{
					LineNumber:  lineNumber,
					BlockNumber: &blockNum,
					IssueType: IssueType{
						Kind: EmptyLineInBlock,
					},
					Severity: Error,
					Context:  &ctx,
				})
				state = betweenBlocks
			} else {
				if !strings.Contains(line, " --> ") {
					blockNum := currentBlock
					issues = append(issues, ValidationIssue{
						LineNumber:  lineNumber,
						BlockNumber: &blockNum,
						IssueType: IssueType{
							Kind:   MalformedTimestamp,
							Value:  line,
							Reason: "missing ' --> ' separator",
						},
						Severity: Error,
						Context:  nil,
					})
				} else {
					parts := strings.Split(line, " --> ")
					if len(parts) == 2 {
						startMs, startErr := s.parseTimestamp(parts[0])
						endMs, endErr := s.parseTimestamp(parts[1])
						switch {
						case startErr == nil && endErr == nil:
							if endMs <= startMs {
								blockNum := currentBlock
								issues = append(issues, ValidationIssue{
									LineNumber:  lineNumber,
									BlockNumber: &blockNum,
									IssueType: IssueType{
										Kind:  InvalidTimestamp,
										Start: parts[0],
										End:   parts[1],
									},
									Severity: Error,
									Context:  nil,
								})
							}
						case startErr != nil:
							blockNum := currentBlock
							issues = append(issues, ValidationIssue{
								LineNumber:  lineNumber,
								BlockNumber: &blockNum,
								IssueType: IssueType{
									Kind:   MalformedTimestamp,
									Value:  parts[0],
									Reason: startErr.Error(),
								},
								Severity: Error,
								Context:  nil,
							})
						default:
							blockNum := currentBlock
							issues = append(issues, ValidationIssue{
								LineNumber:  lineNumber,
								BlockNumber: &blockNum,
								IssueType: IssueType{
									Kind:   MalformedTimestamp,
									Value:  parts[1],
									Reason: endErr.Error(),
								},
								Severity: Error,
								Context:  nil,
							})
						}
					}
				}
				state = expectingText
			}
		case expectingText:
			if isEmpty {
				if !textCollected {
					blockNum := currentBlock
					ctx := "no text before empty line"
					issues = append(issues, ValidationIssue{
						LineNumber:  lineNumber,
						BlockNumber: &blockNum,
						IssueType: IssueType{
							Kind: EmptyText,
						},
						Severity: Error,
						Context:  &ctx,
					})
				}
				state = betweenBlocks
				emptyLineCount = 0
			} else {
				textCollected = true
			}
		}
	}

	if state == betweenBlocks && emptyLineCount > 1 {
		ctx := "trailing empty lines"
		issues = append(issues, ValidationIssue{
			LineNumber:  consecutiveEmptyStart,
			BlockNumber: nil,
			IssueType: IssueType{
				Kind:  MultipleEmptyLines,
				Count: emptyLineCount,
			},
			Severity: Warning,
			Context:  &ctx,
		})
	}

	slog.Debug(fmt.Sprintf("detected problematic lines: %d", len(issues)))

	subtitleService := subtitle.NewSubRipService(s.baseDir)
	_, getAllErr := subtitleService.GetAll(filename)
	isParsable := getAllErr == nil

	errorCount := 0
	warningCount := 0
	for _, i := range issues {
		if i.Severity == Error {
			errorCount++
		}
		if i.Severity == Warning {
			warningCount++
		}
	}
	slog.Info(fmt.Sprintf("diagnostics completed: found %d errors, %d warnings", errorCount, warningCount))

	return DiagnosticReport{
		FilePath:    filePath,
		TotalLines:  totalLines,
		TotalBlocks: currentBlock,
		Issues:      issues,
		IsParsable:  isParsable,
	}, nil
}

// Fix fixes issues in a subtitle file.
// Was SubRipDoctorService::fix.
func (s *SubRipDoctorService) Fix(filename string) (FixReport, error) {
	slog.Info(fmt.Sprintf("starting file fix: %s", filename))

	filePath, err := s.getFilePath(filename)
	if err != nil {
		return FixReport{}, err
	}

	content, err := os.ReadFile(filePath)
	if err != nil {
		return FixReport{}, NewIoError(err)
	}

	lines := splitLines(string(content))

	var outputLines []string
	state := betweenBlocks
	var currentBlockLines []string
	emptyLineCount := 0
	issuesFixed := 0
	var unfixableIssues []ValidationIssue

	for _, line := range lines {
		isEmpty := rustTrim(line) == ""

		switch state {
		case betweenBlocks:
			if isEmpty {
				emptyLineCount++
			} else {
				if len(outputLines) != 0 {
					if emptyLineCount > 1 {
						issuesFixed++
					}
					outputLines = append(outputLines, "")
				}
				emptyLineCount = 0
				currentBlockLines = currentBlockLines[:0]
				currentBlockLines = append(currentBlockLines, line)
				state = expectingTimestamp
			}
		case expectingTimestamp:
			if isEmpty {
				issuesFixed++
			} else {
				currentBlockLines = append(currentBlockLines, line)
				state = expectingText
			}
		case expectingText:
			if isEmpty {
				// End of block - write collected lines
				outputLines = append(outputLines, currentBlockLines...)
				state = betweenBlocks
				emptyLineCount = 0
			} else {
				currentBlockLines = append(currentBlockLines, line)
			}
		}
	}

	if len(currentBlockLines) != 0 {
		outputLines = append(outputLines, currentBlockLines...)
	}

	fixedContent := strings.Join(outputLines, "\n")
	if err := os.WriteFile(filePath, []byte(fixedContent), 0o644); err != nil {
		return FixReport{}, NewIoError(err)
	}

	slog.Debug("fixed file written")

	subtitleService := subtitle.NewSubRipService(s.baseDir)
	_, getAllErr := subtitleService.GetAll(filename)
	validationSuccess := getAllErr == nil

	if validationSuccess {
		slog.Info("validation successful: file can be parsed")
	} else {
		slog.Error("validation failed: file still contains errors")
	}

	slog.Info(fmt.Sprintf("fix completed: %d issues fixed", issuesFixed))

	return FixReport{
		OriginalPath:      filePath,
		FixedPath:         filePath,
		IssuesFixed:       issuesFixed,
		IssuesUnfixable:   len(unfixableIssues),
		UnfixableIssues:   unfixableIssues,
		ValidationSuccess: validationSuccess,
	}, nil
}

// splitOnSeparators splits s on any of the given separator runes, mirroring
// Rust's `str::split(&[char])`. Consecutive separators yield empty fields and
// leading/trailing separators yield empty fields too.
func splitOnSeparators(s string, seps ...rune) []string {
	isSep := func(r rune) bool {
		for _, sep := range seps {
			if r == sep {
				return true
			}
		}
		return false
	}

	var parts []string
	start := 0
	for i, r := range s {
		if isSep(r) {
			parts = append(parts, s[start:i])
			start = i + len(string(r))
		}
	}
	parts = append(parts, s[start:])
	return parts
}

// parseU32 parses a base-10 u32, matching Rust's `str::parse::<u32>()`.
func parseU32(s string) (uint32, error) {
	v, err := strconv.ParseUint(s, 10, 32)
	if err != nil {
		return 0, err
	}
	return uint32(v), nil
}

// splitLines mirrors Rust's str::lines(): splits on '\n', strips an optional
// trailing '\r' from each line, and yields no trailing empty element for a
// terminating '\n'. An empty string yields no lines.
func splitLines(s string) []string {
	if s == "" {
		return nil
	}
	var lines []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			line := s[start:i]
			if len(line) > 0 && line[len(line)-1] == '\r' {
				line = line[:len(line)-1]
			}
			lines = append(lines, line)
			start = i + 1
		}
	}
	if start < len(s) {
		line := s[start:]
		if len(line) > 0 && line[len(line)-1] == '\r' {
			line = line[:len(line)-1]
		}
		lines = append(lines, line)
	}
	return lines
}

// rustTrim strips leading and trailing Unicode whitespace, mirroring Rust's
// str::trim().
func rustTrim(s string) string {
	return strings.TrimFunc(s, isRustWhitespace)
}

// pathExists reports whether the path exists (mirrors Path::exists()).
func pathExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

// isRustWhitespace reports whether r is whitespace per Rust's char::is_whitespace
// (Unicode White_Space).
func isRustWhitespace(r rune) bool {
	switch r {
	case '\t', '\n', '\v', '\f', '\r', ' ', 0x85, 0xA0:
		return true
	}
	switch r {
	case 0x1680, 0x2000, 0x2001, 0x2002, 0x2003, 0x2004, 0x2005, 0x2006,
		0x2007, 0x2008, 0x2009, 0x200A, 0x2028, 0x2029, 0x202F, 0x205F, 0x3000:
		return true
	}
	return false
}
